libs/ui/forms/src/lib/graph-editor.tsx

```diff
@@ -0,0 +1,196 @@
+import ReactFlow, {
+	Background,
+	Controls,
+	Edge,
+	Node,
+	MarkerType,
+	MiniMap,
+	Position
+} from "reactflow";
+import dagre from "@dagrejs/dagre";
+import { Mesh } from "@self-learning/types";
+
+interface ReactFlowGraph {
+	nodes: Node[];
+	edges: Edge[];
+}
+
+const xAxisStartPosition = 20;
+const yAxisStartPosition = 20;
+const xAxisNodeDistance = 260;
+const yAxisDistance = 86;
+const nodeWidth = 172;
+const nodeHeight = 36;
+const platformColor = "#10b981";
+
+function isEdgeInArray(edge: Edge, edges: Edge[]) {
+	for (const elem of edges) {
+		if (edge.label === elem.label) {
+			if (edge.source === elem.source && edge.target === elem.target) {
+				return true;
+			}
+		}
+	}
+	return false;
+}
+
+function removeDuplicateEdges(edges: Edge[]) {
+	const result: Edge[] = [];
+	for (const edge of edges) {
+		if (!isEdgeInArray(edge, result) && edge.target !== "") {
+			result.push(edge);
+		}
+	}
+	return result;
+}
+
+function setEdgeIds(edges: Edge[]) {
+	const result: Edge[] = [];
+	let id = 1;
+	for (const edge of edges) {
+		edge.id = String(id);
+		result.push(edge);
+		id++;
+	}
+	return result;
+}
```

Comment:
skill-lib wird in der nächsten version ein duplicate-remover haben, das sollten wir im kopf behalten. 

Generell kann man hier überlege ob man mit Sets arbeitet um auch einfach zu schauen, ob ein element schon vorhanden ist. 

https://github.com/e-Learning-by-SSE/nm-skill-lib/blob/rf/21-composite-units/src/lib/util/duplicate-remover/duplicate-remover.ts

in dem dortigen IdSet wäre es dann etwas wie 
```
const ed = new IdSet(edges.map(e => ( { ...e, id: e.label )})
ed.has(edge)
```

man muss nur drauf achten dass die edges ein id prop haben